File: Medium/1376-Time Needed to Inform All Employees/DFS.rb
# @param {Integer} n
# @param {Integer} head_id
# @param {Integer[]} manager
# @param {Integer[]} inform_time
# @return {Integer}
def num_of_minutes(_n, head_id, manager, inform_time)
  subordinates = {}

  (0...manager.size).each do |i|
    subordinates[manager[i]] = [] unless subordinates.member?(manager[i])
    subordinates[manager[i]].push(i)
  end

  dfs(head_id, subordinates, inform_time)
end

# @param {Integer} head_id
# @param {Hash} subordinates
# @param {Integer[]} inform_time
# @return {Integer}
def dfs(head_id, subordinates, inform_time)
  return 0 unless subordinates.member?(head_id)

  inform_time[head_id] +
    subordinates[head_id].map { |id| dfs(id, subordinates, inform_time) }.max
end

File: Problemset/1376-Time Needed to Inform All Employees/DFS.rs
use std::collections::HashMap;

impl Solution {
    pub fn num_of_minutes(n: i32, head_id: i32, manager: Vec<i32>, inform_time: Vec<i32>) -> i32 {
        let mut subordinates = HashMap::new();

        for i in 0..manager.len() {
            subordinates
                .entry(manager[i])
                .or_insert(vec![])
                .push(i as i32);
        }

        Self::dfs(head_id, &subordinates, &inform_time)
    }

    fn dfs(head_id: i32, subordinates: &HashMap<i32, Vec<i32>>, inform_time: &Vec<i32>) -> i32 {
        match subordinates.get(&head_id) {
            Some(subs) => {
                inform_time[head_id as usize]
                    + subs
                        .iter()
                        .map(|&id| Self::dfs(id, subordinates, inform_time))
                        .max()
                        .unwrap()
            }
            None => 0,
        }
    }
}
